File: .claude/skills/agent-skill-kit-commons/packages/py/src/skills_ref/rules/references.py
from __future__ import annotations

import re
from pathlib import Path

from ..types import Issue, Parsed

_LINK_RE = re.compile(r"\[[^\]]*\]\(([^)#?]+)(?:[?#][^)]*)?\)")
_URL_RE = re.compile(r"^(https?:|mailto:|tel:|ftp:|#)")


def check_references(parsed: Parsed, skill_dir: str) -> list[Issue]:
    issues: list[Issue] = []
    body = parsed.body or ""
    base = Path(skill_dir)
    for m in _LINK_RE.finditer(body):
        target = m.group(1).strip()
        if not target:
            continue
        if _URL_RE.match(target):
            continue
        if target.startswith("/"):
            continue
        full = (base / target).resolve()
        if not full.exists():
            issues.append(
                Issue(
                    "W002_BROKEN_REFERENCE",
                    "warn",
                    "body",
                    f"broken reference: {target}",
                )
            )
    return issues

File: .claude/skills/agent-skill-kit-commons/packages/mjs/src/rules/references.mjs
import { existsSync, statSync } from "node:fs";
import { resolve } from "node:path";
import { makeIssue } from "../types.mjs";

const LINK_RE = /\[[^\]]*\]\(([^)#?]+)(?:[?#][^)]*)?\)/g;

export function checkReferences(parsed, skillDir) {
  const issues = [];
  const body = parsed.body ?? "";
  for (const m of body.matchAll(LINK_RE)) {
    const target = m[1].trim();
    if (!target) continue;
    if (/^(https?:|mailto:|tel:|ftp:|#)/.test(target)) continue;
    if (target.startsWith("/")) continue; // absolute paths handled by checkBody (E010)
    const full = resolve(skillDir, target);
    try {
      if (!existsSync(full)) {
        issues.push(
          makeIssue(
            "W002_BROKEN_REFERENCE",
            "warn",
            "body",
            `broken reference: ${target}`,
          ),
        );
      } else {
        statSync(full);
      }
    } catch {
      issues.push(
        makeIssue(
          "W002_BROKEN_REFERENCE",
          "warn",
          "body",
          `broken reference: ${target}`,
        ),
      );
    }
  }
  return issues;
}
